src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -15,9 +15,95 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
     /// A simple class that combines a single type symbol with annotations
     /// </summary>
     [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
-    internal abstract class TypeSymbolWithAnnotations : IFormattable
+    internal readonly struct TypeSymbolWithAnnotations : IFormattable
     {
-        public sealed override string ToString() => TypeSymbol.ToString();
+        /// <summary>
+        /// A builder for lazy instances of TypeSymbolWithAnnotations.
+        /// </summary>
+        [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+        internal struct Builder
+        {
+            private TypeSymbol _defaultType;
+            private INonNullTypesContext _nonNullTypesContext;
+            private bool _isAnnotated;
+            private Extensions _extensions;
+
+            /// <summary>
+            /// The underlying type, unless overridden by _extensions.
+            /// </summary>
+            internal TypeSymbol DefaultType => _defaultType;
+
+            /// <summary>
+            /// True if the fields of the builder are unset.
+            /// </summary>
+            internal bool IsNull => _defaultType is null;
+
+            /// <summary>
+            /// Set the fields of the builder.
+            /// </summary>
+            internal bool InterlockedInitialize(TypeSymbolWithAnnotations type)
+            {
+                if ((object)_defaultType != null)
+                {
+                    return false;
+                }
+                _isAnnotated = type.IsAnnotated;
+                Interlocked.CompareExchange(ref _nonNullTypesContext, type.NonNullTypesContext, null);
+                Interlocked.CompareExchange(ref _extensions, type._extensions, null);
+                return (object)Interlocked.CompareExchange(ref _defaultType, type._defaultType, null) == null;
+            }
+
+            /// <summary>
+            /// Create immutable TypeSymbolWithAnnotations instance.
+            /// </summary>
+            internal TypeSymbolWithAnnotations ToType()
+            {
+                return (object)_defaultType == null ?
+                    default :
+                    new TypeSymbolWithAnnotations(_defaultType, _nonNullTypesContext, _isAnnotated, _extensions);
+            }
+
+            internal string GetDebuggerDisplay() => ToType().GetDebuggerDisplay();
+        }
+
+        /// <summary>
+        /// The underlying type, unless overridden by _extensions.
+        /// </summary>
+        private readonly TypeSymbol _defaultType;
+
+        /// <summary>
+        /// Additional data or behavior. Such cases should be
+        /// uncommon to minimize allocations.
+        /// </summary>
+        private readonly Extensions _extensions;
+
+        /// <summary>
+        /// Returns:
+        /// false for string, int, T;
```

Comment:
We may want a different field than `IsAnnotated` in that case, or a renamed field, since `T` was unannotated. Let's revisit this comment then.

---
In reply to: [207051401](https://github.com/dotnet/roslyn/pull/28943#discussion_r207051401) [](ancestors = 207051401)